### PR TITLE
PF-522: Use StairwayComponent in WSM

### PR DESCRIPTION
**Pull request by jaycarlton**
Split the stairway initialization out of JobService and gave it its own StairwayComponent.

I'm initializing stairway in the startup initializer now, the same way I'm doing for buffer service.

### CODE REVIEW

File: src/main/java/bio/terra/workspace/service/job/JobService.java
Comment (jaycarlton) [inline src/main/java/bio/terra/workspace/service/job/JobService.java:262]: should the stairway component just expose these functions instead of exposing the stairway object with get()?

File: src/main/java/bio/terra/workspace/app/Main.java
Comment (yonghaoy) [inline src/main/java/bio/terra/workspace/app/Main.java:26]: nit:order by alphabetical

File: src/main/java/bio/terra/workspace/service/job/JobService.java
Comment (yonghaoy) [reply src/main/java/bio/terra/workspace/service/job/JobService.java:262]: I don't suggest that. Feels we are putting to much into stairwayComponent, and make it become part of 'Stairway library' which is unnecessary. And if Stairway method changes, we have to upgrade stairwayComponent correspondingly. 
Using stairwayComponent just for initialization and health status should be good enough.

File: src/main/java/bio/terra/workspace/service/job/JobService.java
Comment (yonghaoy) [inline src/main/java/bio/terra/workspace/service/job/JobService.java:80]: then you lose StairwayExceptionSerializer. 
You may have to change method common-lib's initialize method. (maybe passing Stairway.Builder() if we are seeing too many parameters)?
@ddietterich wdyt?

File: src/main/java/bio/terra/workspace/service/job/JobService.java
Comment (jaycarlton) [reply src/main/java/bio/terra/workspace/service/job/JobService.java:262]: 👍🏻

File: src/main/java/bio/terra/workspace/service/job/JobService.java
Comment (ddietterich) [reply src/main/java/bio/terra/workspace/service/job/JobService.java:80]: We definitely need a way to supply an exception serializer and override the Stairway built-in one.
If the param list gets too long, we can make an initializeStairway object with a builder and pass that in.

File: src/main/java/bio/terra/workspace/service/job/JobService.java
Comment (ddietterich) [reply src/main/java/bio/terra/workspace/service/job/JobService.java:262]: I agree with @yonghaoy.
We could capture the Stairway object into a `private final Stairway stairway` from the stairwayComponent in the JobService constructor and keep using the `stairway` variable, right?

File: src/main/resources/application.yml
Comment (ddietterich) [inline src/main/resources/application.yml:84]: Probably more than we need. I'd suggest 50. It is per node and we probably don't want more than 150 threads competing for access to the backend database server.

**Review by ddietterich — COMMENTED** [review]
This is looking good. I think we need a solution for the exception serialization before we callit done.

File: src/main/java/bio/terra/workspace/service/job/JobService.java
Comment (jaycarlton) [reply src/main/java/bio/terra/workspace/service/job/JobService.java:262]: You can't initialize a final field except in the constructor, or at least it's a warning.

I'm concerned we don't have a clean abstraction. Either `StairwayComponent` owns `Stairway`, or it doesn't. I don't like `get()`, but I like saving the result of a `get()` even less. It doesn't _belong_ to `JobService`.

Should that method be removed from `StairwayComponent`, and we just return a `Stairway` from the initialize method so it's like a factory? That approach wouldn't make sense with the status, unless we assume `Stairway` really is a singleton.

File: src/main/java/bio/terra/workspace/service/job/JobService.java
Comment (jaycarlton) [reply src/main/java/bio/terra/workspace/service/job/JobService.java:80]: Do we need this before merging this (and/or before the release)?

Comment (ddietterich) [reply src/main/java/bio/terra/workspace/service/job/JobService.java:80]: The exception serializer improves the fidelity of exceptions caught with Stairway. But the default one in Stairway will deliver messages. So I vote not to let this hold up the merge. We can do it as a follow-on.

Let's discuss separately whether we can have one shared serializer for MC Terra based on the common exceptions or if we need to allow for individual ones per component.

File: src/test/resources/application-test.yml
Comment (yonghaoy) [reply src/test/resources/application-test.yml:36]: Buffer test does not have them. Those are from Buffer's in main's `application.yml`, not `test`.
For test, the only config Buffer uses is to reset stairway. And that is `force-clean-start: true` in test, but `force-clean-start: false` in prod. I do see WSM has similar configs here:
https://github.com/DataBiosphere/terra-workspace-manager/blob/dev/src/main/java/bio/terra/workspace/app/configuration/external/StairwayDatabaseConfiguration.java#L18, and seems WSM does not reset Stairway when local/test server starts, but it does upgrade (see this file's line 19: migrate-upgrade: true), now it should be replaced by your new config dedpends on if you want to use this

Comment (ddietterich) [reply src/test/resources/application-test.yml:36]: Those setting look good to me. I think we should be doing `force-clean-start: true` on our test runs.

**Review by ddietterich — APPROVED** [review]
lgtm

File: src/main/java/bio/terra/workspace/service/job/JobService.java
Comment (jaycarlton) [reply src/main/java/bio/terra/workspace/service/job/JobService.java:262]: Sorry if that sounded aggresive.

File: src/test/resources/application-test.yml
Comment (yonghaoy) [reply src/test/resources/application-test.yml:36]: This does not look correct to me. At least we should remove the migrate-upgrade in line 19? I don't know which config you are reading.

Comment (yonghaoy) [reply src/test/resources/application-test.yml:36]: And you should remove all lin 38-42 from here and put them in application.yml not in test

**Review by yonghaoy — APPROVED** [review]
Thanks

File: src/main/resources/application.yml
Comment (yonghaoy) [inline src/main/resources/application.yml:85]: migrate-upgrade should be true in non-testing env. force-clean-start should be `${env.db.init}`
See line 59 and 60 above, and you should remove those two lines because they are not used anymore. 
Also those configs should be deleted as well:
https://github.com/DataBiosphere/terra-workspace-manager/blob/dev/src/main/java/bio/terra/workspace/app/configuration/external/StairwayDatabaseConfiguration.java#L13-L42
Those two are replaced by your new code: https://github.com/DataBiosphere/terra-common-lib/blob/develop/src/main/java/bio/terra/common/stairway/StairwayProperties.java#L19-L24

**Pull request merged**
